Userland/Libraries/LibCrypto/Cipher/ChaCha20.cpp

```diff
@@ -0,0 +1,159 @@
+void ChaCha20::process_block()
+{
+    // Copy the current state into the block
+    memcpy(m_state.block, m_state.state, 16 * sizeof(u32));
+
+    // ChaCha20 runs 20 rounds, alternating between "column rounds" and "diagonal rounds".
+    // Each round consists of four quarter-rounds
+    for (u32 i = 0; i < 20; i += 2) {
+        // Column rounds
+        do_quarter_round(m_state.block[0], m_state.block[4], m_state.block[8], m_state.block[12]);
+        do_quarter_round(m_state.block[1], m_state.block[5], m_state.block[9], m_state.block[13]);
+        do_quarter_round(m_state.block[2], m_state.block[6], m_state.block[10], m_state.block[14]);
+        do_quarter_round(m_state.block[3], m_state.block[7], m_state.block[11], m_state.block[15]);
+
+        // Diagonal rounds
+        do_quarter_round(m_state.block[0], m_state.block[5], m_state.block[10], m_state.block[15]);
+        do_quarter_round(m_state.block[1], m_state.block[6], m_state.block[11], m_state.block[12]);
+        do_quarter_round(m_state.block[2], m_state.block[7], m_state.block[8], m_state.block[13]);
+        do_quarter_round(m_state.block[3], m_state.block[4], m_state.block[9], m_state.block[14]);
+    }
+
+    // At the end of 20 rounds, we add the original input words to the output words,
+    for (u32 i = 0; i < 16; i++) {
+        m_state.block[i] += m_state.state[i];
+    }
+
+    // and serialize the result by sequencing the words one-by-one in little-endian order.
+    for (u32 i = 0; i < 16; i++) {
+        m_state.block[i] = AK::convert_between_host_and_little_endian(m_state.block[i]);
+    }
+}
+
+ALWAYS_INLINE static void rotl(u32& x, u32 n)
+{
+    x = (x << n) | (x >> (32 - n));
+}
+
+// https://datatracker.ietf.org/doc/html/rfc8439#section-2.1
+void ChaCha20::do_quarter_round(u32& a, u32& b, u32& c, u32& d)
+{
+    a += b;
+    d ^= a;
+    rotl(d, 16);
+
+    c += d;
+    b ^= c;
+    rotl(b, 12);
+
+    a += b;
+    d ^= a;
+    rotl(d, 8);
+
+    c += d;
+    b ^= c;
+    rotl(b, 7);
+}
+
+void ChaCha20::run_cipher(ReadonlyBytes input, Bytes& output)
+{
+    size_t offset = 0;
+    while (offset < input.size()) {
+        if (m_state.block_offset == 0 || m_state.block_offset >= 64) {
+            // Generate a new XOR block
+            process_block();
+
+            // Increment the block counter, and carry over to block 13
+            m_state.state[12]++;
+            if (m_state.state[12] == 0) {
+                m_state.state[13]++;
+            }
+
+            m_state.block_offset = 0;
+        }
+
+        // XOR the input and the current block
+        u32 n = min(input.size() - offset, 64 - m_state.block_offset);
+        u8* key_block = (u8*)m_state.block + m_state.block_offset;
+        for (u32 i = 0; i < n; i++) {
+            u8 input_byte = input.offset_pointer(offset)[i];
+            u8 key_byte = key_block[i];
+
+            ByteReader::store(output.offset_pointer(offset + i), input_byte ^ key_byte);
```

Comment:
ASAN is complaining here that your offset + i is exactly the same as the sizeof(output), and you're writing one outside the bounds of the ByteBuffer you created in the test.